### DISCUSSION TITLE
Skeleton v3 (Next) - Proposal, RFC, and Updates

### DISCUSSION THREAD

**Discussion by endigo9740** [skeletonlabs/skeleton#2375] [Announcements]
![image](https://github.com/skeletonlabs/skeleton/assets/1509726/ff217587-6514-4adb-af42-c371065f3432)

Hey everyone, Chris here from the Skeleton crew. To help celebrate the new year, the team and I wanted to take a moment of your time to share some exciting details about the next major version of Skeleton.

In short, we’ve decided that version 3.x of Skeleton will represent a much larger step forward than any previous release to date. In fact, over the last few months, the team and I have internally dubbed this “Skeleton Next” - as in the “next major version of Skeleton”. In order to help convey the scope and scale of these changes, we’ve shared a detailed proposal below covering the scope of changes:

- Proposal: [View the full proposal](https://www.notion.so/Skeleton-Next-Proposal-56861fb914ec474c98003c29296a1d5f?pvs=21)
-  Doc Site: https://next.skeleton.dev/

## Progress Updates

Moved to the [Pre-Release announcement post](https://github.com/skeletonlabs/skeleton/discussions/2681).

## Video Walkthrough

[![image](https://github.com/skeletonlabs/skeleton/assets/1509726/b278ed0a-1e77-458f-8fb3-6ac1d2c1bcf4)](https://www.youtube.com/watch?v=auFvaLyN9PQ)

## Summary of Changes

Again, we strongly recommend you read the proposal to understand the extent of the updates. However, let's review some of the most exciting changes:

- Skeleton Next will introduce a new and more modular architecture to Skeleton. Alongside this, we will move beyond Svelte, becoming fully framework agnostic. The first iteration will see support added for [React](https://react.dev/), with additional frontend frameworks introduced over time.
- We’ve been closely monitoring the progress of [Svelte 5](https://svelte-5-preview.vercel.app/docs/introduction), and overall we feel it represents a huge step forward. As part of Skeleton Next we’ll plan to update and rewrite many portions of Skeleton to take advantage of new features like [Runes](https://svelte.dev/blog/runes), [Snippets](https://svelte-5-preview.vercel.app/docs/snippets), and [more](https://component-party.dev/?f=svelte4,svelte5).
- We will continue to provide deep integration with Tailwind and it’s latest features, while also monitoring notable updates such as [Tailwind Oxide](https://www.youtube.com/watch?v=CLkxRnRQtDE&t=2146s&themeRefresh=1).
- We feel one of the best features of Skeleton today is our adaptive theme system, made possible through [atomic design](https://atomicdesign.bradfrost.com/) and [design tokens](https://www.youtube.com/watch?v=JbxKTBvSLeY&t=1s). Next will introduce first major expansion to themes since launch. Including a new standalone theme generator.
- We know many companies and professionals rely on design tools such as [Figma](http://figma.com/). In order to help help get you up and running quickly, we will be introducing our first official Figma design assets.
- In order to properly document and support multiple frameworks running in parallel, we will be migrating our documentation website to utilize [Astro](https://astro.build/). This provides a number benefits out of the box, including: markdown support, i18n, and more.
- Please note that we will continue to support meta frameworks like [SvelteKit v2.x](https://kit.svelte.dev/), and even expand to new options, such as [Next.js for React](https://nextjs.org/).
- Finally, we will be aiming to provide additional tooling to support developers, including major updates to the Skeleton CLI, and (potentially) a new VS Code extension.

## FAQ

- Q: Does this mean work on Skeleton v2 will be halted?
    - A: No. We encourage you to continue sending PRs for bug fixes and minor feature updates. The Skeleton team will also continue to provide support for v2. However, internally we will be shifting our focus to v3. Especially as Svelte 5 reaches maturity.
- Q: Is `{insert feature}` coming in Skeleton v3?
    - A: Check the [v3 milestones on GitHub](https://github.com/skeletonlabs/skeleton/milestone/5). Feedback and feature requests are encouraged!
- Q: When will an initial version of Next be available to preview and test?
    - A: Aside from the early prototype shared in the proposal, we've already started work on a new `v3` branch of Skeleton. Expect periodic updates as we hit certain milestones and features and documentation are flushed out.
- Q: When can we expect a Skeleton v3 to launch?
    - A: Portions of this initiative will be spread across 2024. We anticipate the React module will arrive first, with Svelte 5 soon after. Pre-release versions of each of those will be announced closer to launch.
- Q: How can I contribute to these new updates and features?
    - A: Keep and eye on GitHub, Discord, and our social channels. We’ll put out a call to arms when we’re ready for more volunteers. The goal will be to establish a solid foundation internally, then introduce more contributors over time.
- Q: I’m interested in creating a port to `{some other framework}`. When can I start?
    - A: We highly encourage folks to pace themselves until the base design patterns and paradigms have been established for Svelte and React. However, it never hurts to start planning early!
- Q: How can I volunteer as a Community Advisor?
    - A: Contact information has been provided in the dedicated section near the bottom of the proposal. We’re actively seeking React and Figma experts asap!

## Where can I submit comments?

The whole purpose of an RFC is to gather comments from folks like you, our wonderful community. We encourage you to review our proposal and FAQ in full. Then please leave your feedback in one of the two following locations:

1. Reach out in the comments section below.
2. The new `#skeleton-next` channel on Skeleton's [Discord](https://discord.gg/EXqV7W8MtY).

## Let's keep it constructive

Please note that Skeleton does provide and enforce a [Code of Conduct](https://github.com/skeletonlabs/community/blob/main/CODE_OF_CONDUCT.md). We take this very seriously. While we understand that major changes and be disruptive and frustrating for some users, we please ask you to keep discourse civil and constructive. Any users being rude or abrasive towards our team or other members of the community will be subject to moderation. That said, we're very excited about the possibilities this unlocks, and really want to hear what you think. Please do share!

**Comment by zadamg** [comment]
amazing. 

the design system is already awesome and has made life so much easier for me. excited to see all the other possibilities with runes, shallow routing, and the standalone theme generator is a big step as well. 

👏

**Comment by khionu** [comment]
> As part of Skeleton Next we’ll plan to update and rewrite many portions of Skeleton to take advantage of new features like [Runes](https://svelte.dev/blog/runes), [Snippets](https://svelte.dev/blog/runes), and [more](https://component-party.dev/?f=svelte4,svelte5).

Fyi, the Snippets link points to Runes

**Reply by endigo9740** [reply]
Fixed, thank you!

**Comment by codercatdev** [comment]
I have a slight concern that Skeleton will focus on working for all the frameworks and not be something I love that specifically works well with Svelte. With that said I love the product and the team that surrounds it so here are just a few things to think about.

## Agnostic

Something that we might want to consider moving forward is that to be truly framework agnostic we wouldn't want to rely as much on the framework's specific tools like including Runes directly into the codebase. 

We could instead take the approach like [Ionic](https://ionicframework.com/) did with [Stencil](https://stenciljs.com/). 

If you want to go the route of creating framework specific components like [Papanasi](https://github.com/CKGrafico/papanasi) did with [Mitosis](https://github.com/BuilderIO/mitosis) this would be another route. This would allow us to tap into the work being done with the Figma plugin that Builder has created as well.

## Cross Platform State

should we consider anything around state?

Chakra UI went through some pain from a state aspect and built [Zag.js](https://zagjs.com/overview/introduction#motivation). 

There are other frameworks like [Xoid](https://www.xoid.dev/)

**Reply by endigo9740** [reply]
Great feedback Alex, I would like to think of this as more of embracing the best of each framework, rather than aiming for the lowest common denominator. That's why we're very intentionally introducing new frameworks very slowly over time. With only the addition of React coming in v3.

To address a few concerns you have specifically:

1. The core package in Skeleton `@skeletonlabs/skeleton` will have a single dependency: Tailwind. If you want to use our theme system, design tokens, element styles, etc. Then all of that will be truly agnostic. Wether you're using a component framework or not.
2. For each framework, think of the mental model as a "plugin". So we'll have a plugin that adds Svelte-specific components, a plugin that adds React-specific components. You bring those in only when you need them. And all of these are designed to be plug and play alongside the core features, enabling the all the same features folks know and love in Skeleton's components today (themes, style props, etc)
3. We're VERY specifically building the Svelte package `@skeletonlabs/skeleton-svelte` around Svelte 5. Which means we won't have an intermittent migration phase from Svelte 4 -> 5 specifically. Rather, Skeleton v3 will default to Svelte 5 as the baseline. Which means we're writing all of our Svelte features with Runes, Snippets, and any other Svelte 5 goodies right away!
4. Per state - there will be exceptions (like modals/toasts/drawers), but almost none of our components today utilize framework specific state management. For Svelte 5 we'll likely use `$state()` runes. For React, we'll likely use the Context API + Provider model. In the React world alone there's waaaayyyy too many state options to take an opinionated stance without alienating a large portion of our audience. So as much as possible we're going to be state library agnostic.

Thank you for sharing the links to those other libraries though. We have definitely looked at several of those already, and will keep the others in mind going forward.

Also I would very highly recommend reviewing our prototype. Which shows how we're implementing components between Svelte and React. The common conventions, and where things diverge:
https://github.com/skeletonlabs/skeleton-next

Note that I have plans to release another progress update for "Next" alongside the new release this upcoming Tuesday. That will cover two topics:

1. Design aesthetic updates
2. The new code style guidelines for authoring components.

The latter provides a great breakdown of how we're aiming to keep standard conventions and paradigms between multiple frameworks. I'm very pleased with where this currently stands. But as always feedback is welcome!

**Reply by endigo9740** [reply]
@codercatdev those updates I mentioned above are now available here:

- https://github.com/skeletonlabs/skeleton/discussions/2424

**Comment by jhechtf** [comment]
My one nitpick that I've got with skeleton is the system for things like toasts and drawers. I personally do not like having to use a singular thing for drawers, and would appreciate if we could revisit how those are setup.

**Reply by endigo9740** [reply]
Hey @jhechtf, can you explain what you mean? Do you mean the singleton pattern - where's there's a single/reused instance of the Drawer or Toast component per application? I don't think that'll change going forward, but we do have some ideas to simplify the process. Right now there's kind of choice overload, and provides a lot of complexity - both within the features and for end users trying to learn how this works.

Conceptionally though, we're confident this is the most scalable option. But as with most features in Skeleton, it will remain opt-in. Which means if you don't like the way we handle them, you can easily bring in an alternative and still utilize your Skeleton styles and themes in most cases.

**Reply by jhechtf** [reply]
I personally do not like having to use a store to show a drawer, especially since I use them fairly heavily in my application. In the current setup I'd have to have huge if/else blocks. Some of the drawers are very specific in what they do, and it makes more sense for those to be their own thing that are controlled outside of the current singular store.

That, by far, has been my biggest issue with Skeleton so far. I created a drawer component myself by ~~stealing~~ borrowing a lot of the markup. I think for toasts it makes more sense, I still am personally not a fan of it, but I truly believe that the drawers should not be singletons.

**Reply by Sarenor** [reply]
So if the Drawers work like Modals where you can either pass in a component reference or pre-register specific views with specific components, it would work for your usecase?

**Reply by jhechtf** [reply]
I'm not huge on the way the modals currently work. I would really rather just be able to 

```svelte
<script>
import { Drawer } from '@skeleton/pkg';
let open = false;
</script>

<button on:click={() => open = true }>Open</button>

<Drawer {open}>
 <!-- W/E here -->
</Drawer>
```

And personally, the same for Modals as well. It's nice that the modal store can at least accept a component to render instead but I'd really rather not have to use the store setup at all.

**Reply by rChaoz** [reply]
I think with Svelte 5 this would be doable - the drawer component would save its `children` prop somewhere globally, where it would then be `{@rendered}` by a top-level `<DrawerManager />` component from the root `+layout.svelte`. Would that work?

**Reply by endigo9740** [reply]
It's still too early to give a definitive answer to how we're going to handle modals/drawers/toasts in Skeleton v3, but we have been noting LOTS of ideas here:

- https://github.com/skeletonlabs/skeleton/issues/2357
- https://github.com/skeletonlabs/skeleton/issues/2397
- https://github.com/skeletonlabs/skeleton/issues/2398

I'd encourage any additional feedback to directed to these threads.

Likewise if you wish to check any other v3 milestone features, you can do so here:

- https://github.com/skeletonlabs/skeleton/milestone/5

**Comment by endigo9740** [comment]
Just sharing in case you missed it everyone - but the new Skeleton v3 pre-release is now available!

- https://github.com/skeletonlabs/skeleton/discussions/2681

**Reply by oneezy** [reply]
Loving it!
